### PR TITLE
Add support for hls, dash, smoothstreaming and caching

### PR DESCRIPTION
**Pull request by gazedash**
Thanks for opening a PR!
Since this is a volunteer project and is very active, anything you can do to reduce the amount of time needed to review and merge your PR is appreciated.
The following steps will help get your PR merged quickly:

#### Update the documentation
If you've added new functionality, update the README.md with an entry for your prop or event.
The entry should be inserted in alphabetical order.

#### Update the changelog
After you open the PR, update the CHANGELOG.md file with an entry pointing to your PR.

#### Provide an example of how to test the change
If the PR requires special testing setup provide all the relevant instructions and files. This may include a sample video file or URL, configuration, or setup steps.

#### Focus the PR on only one area
Testing multiple features takes longer than isolated changes and if there is a bug in one feature, prevents the other parts of your PR from getting merged until it gets fixed.
If you're touching multiple different areas that aren't related, break the changes up into multiple PRs.

#### Describe the changes
Add a note describing what your PR does. If there is a change to the behavior of the code, explain why it needs to be updated.

### CODE REVIEW

**Comment by Yunnathai** [comment]
any update？need it ～

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java [hunk 1]
```diff
@@ -333,6 +371,15 @@ private void reLayout(View view) {
     }
 
     private void initializePlayer() {
+                if(cacheMaxSize > 0) {
```
Comment (danielmarino24i) [inline android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java:374]: Wrong indentation

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java [hunk 2]
```diff
@@ -383,28 +430,63 @@ private void initializePlayer() {
...
-                return new DashMediaSource(uri, buildDataSourceFactory(false),
-                        new DefaultDashChunkSource.Factory(mediaDataSourceFactory), 
-                        minLoadRetryCount, DashMediaSource.DEFAULT_LIVE_PRESENTATION_DELAY_MS,
-                        mainHandler, null);
-            case C.TYPE_HLS:
-                return new HlsMediaSource(uri, mediaDataSourceFactory, 
-                        minLoadRetryCount, mainHandler, null);
-            case C.TYPE_OTHER:
-                return new ExtractorMediaSource(uri, mediaDataSourceFactory, new DefaultExtractorsFactory(),
+            case "smoothstreaming":
```
Comment (danielmarino24i) [inline android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java:474]: Why aren't you using the constants?

**Comment by gazedash** [comment]
@jamninetyfive 
You could really help if you could fix formatting and use constants instead of string literals!
Didn't have time to fix that yet

**Comment by gazedash** [comment]
~~FYI: This PR breaks local file playback, need to fix this~~

Fixed local file playback in the latest commit!

**Comment by Yunnathai** [comment]
@gazedash well done good~ bro

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/Utils.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+import com.facebook.react.bridge.Promise;
+import com.facebook.react.views.imagehelper.ResourceDrawableIdHelper;
+import com.google.android.exoplayer2.upstream.RawResourceDataSource;
+
+/**
+ * @author Guichaguri
+ */
+public class Utils {
+
+    public static final String EVENT_INTENT = "com.guichaguri.trackplayer.event";
```
Comment (cobarx) [inline android-exoplayer/src/main/java/com/brentvatne/exoplayer/Utils.java:19]: This should be namespaced under the com.brentvatne.exoplayer namespace.

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/Utils.java [hunk 4]
```diff
@@ -0,0 +1,148 @@
+package com.brentvatne.exoplayer;
```
Comment (cobarx) [inline android-exoplayer/src/main/java/com/brentvatne/exoplayer/Utils.java:1]: Add a LICENSE tag acknowledging the author of react-native-track-player.

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java (hunk 2 above)
Comment (cobarx) [inline android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java:474]: We should not be scrapping the existing media source creators. They contain lots of configuration options that people need.

**Comment by redreceipt** [comment]
so before this PR is merged, is the only option for Android mp4 files?

**Comment by CHaNGeTe** [comment]
@redreceipt nope, that´s a pretty old PR, hls / dash / ss is supported.
I will check if there is sth interesting here, and if not, will close it.

**Comment by cobarx** [comment]
Seems like this is enabling caching for all the formats listed in the title. Not sure how well it works though. My concern with all these caching PRs is they often force everything to go through the caching code path even if you're not using it.

**Comment by redreceipt** [comment]
I'm seeing an issue where HLS plays on the Sim but not the actual device. Any thoughts to what it could be?

**Comment by redreceipt** [comment]
This is why it seems to not be working, didn't know `exoplayer` wasn't the default

https://github.com/react-native-community/react-native-video/issues/1746

**Comment by RupamShaw** [comment]
Is still this library not support smoothstreaming?

**Comment by devnullpointer** [comment]
not sure why no one is using ExoPlayer's default cache mechanism. This is what I had done on my native android app:

```
public class CacheDataSourceFactory implements DataSource.Factory
{
    private final Context context;
    private final DefaultDataSourceFactory defaultDatasourceFactory;
    private final long maxFileSize;

    private SimpleCache simpleCache;

    public CacheDataSourceFactory(final Context context, final long maxCacheSize, final long maxFileSize, final String[] headers)
    {
        super();

        this.context = context;
        this.maxFileSize = maxFileSize;

        final DefaultBandwidthMeter bandwidthMeter = new DefaultBandwidthMeter.Builder(context).build();
        final DefaultHttpDataSourceFactory factory = new DefaultHttpDataSourceFactory(Config.USER_AGENT, bandwidthMeter);

        for (int idx = 0; idx < headers.length; idx++)
        {
            if (!headers[idx].isEmpty())
            {
                final String[] nameValue = headers[idx].split("=");
                if (nameValue.length > 0)
                {
                    factory.getDefaultRequestProperties().set(nameValue[0], nameValue[1]);
                }
            }
        }

        defaultDatasourceFactory = new DefaultDataSourceFactory(this.context, bandwidthMeter, factory);
        simpleCache = ExoSimpleCache.getInstance(context, maxCacheSize);
    }

    @Override
    public DataSource createDataSource()
    {
        return new CacheDataSource(simpleCache, defaultDatasourceFactory.createDataSource(),
                new FileDataSource(), new CacheDataSink(simpleCache, maxFileSize),
                CacheDataSource.FLAG_BLOCK_ON_CACHE | CacheDataSource.FLAG_IGNORE_CACHE_ON_ERROR, null);
    }
}
```


then in my exoplayer class:
```
private void initializePlayer()
    {
        ... 
        mediaSource = buildMediaSource(playbackUri, playbackUri.getPath().substring(playbackUri.getPath().lastIndexOf(".") + 1));
        mediaSource.addEventListener(mHandler, eventLogger);

        if (srtUri != null)
        {
            // Set subtitles
            final Format textFormat = Format.createTextSampleFormat(null, MimeTypes.APPLICATION_SUBRIP, Format.NO_VALUE, "en");
            final MediaSource textMediaSource = new SingleSampleMediaSource.Factory(mediaDataSourceFactory).createMediaSource(srtUri, textFormat, C.TIME_UNSET);
            final MediaSource mediaSourceWithText = new MergingMediaSource(mediaSource, textMediaSource);
            mediaSourceWithText.addEventListener(mHandler, eventLogger);

            player.prepare(mediaSourceWithText);
        }
        else
        {
            player.prepare(mediaSource);
        }
    }

@Override
    protected void onCreate(@Nullable Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);
        ...
        mediaDataSourceFactory = buildDataSourceFactory();
       ....
    }

private MediaSource buildMediaSource(final Uri uri, final String overrideExtension)
    {
        final int type = TextUtils.isEmpty(overrideExtension) ? Util.inferContentType(uri) : Util.inferContentType("." + overrideExtension);
        switch (type)
        {
            case C.TYPE_SS:
                return new SsMediaSource.Factory(mediaDataSourceFactory).createMediaSource(uri);
            case C.TYPE_DASH:
                return new DashMediaSource.Factory(mediaDataSourceFactory).createMediaSource(uri);
            case C.TYPE_HLS:
                return new HlsMediaSource.Factory(mediaDataSourceFactory).createMediaSource(uri);
            case C.TYPE_OTHER:
                return new ProgressiveMediaSource.Factory(mediaDataSourceFactory).createMediaSource(uri);
            default:
            {
                throw new IllegalStateException("Unsupported type: " + type);
            }
        }
    }

private DataSource.Factory buildDataSourceFactory()
    {
        final CacheDataSourceFactory factory = new CacheDataSourceFactory(this,100 * 1024 * 1024, 5 * 1024 * 1024, headers);
        return new DefaultDataSourceFactory(this, null, factory);
    }

```

**Pull request closed without merging**